tests/test_stormwater.py

```diff
@@ -0,0 +1,769 @@
+import functools
+import math
+import numpy
+import os
+import pandas
+import pygeoprocessing
+import tempfile
+import unittest
+from unittest import mock
+
+from osgeo import gdal, ogr, osr
+
+
+def mock_iterblocks(*args, xoffs=[], xsizes=[], yoffs=[], ysizes=[], **kwargs):
+    """Mock function for pygeoprocessing.iterblocks that yields custom blocks.
+
+    Args:
+        xoffs (list[int]): list of x-offsets for each block in order
+        xsizes (list[int]): list of widths for each block in order
+        yoffs (list[int]): list of y-offsets for each block in order
+        ysizes (list[int]): list of heights for each block in order
+
+    Yields:
+        dictionary with keys 'xoff', 'yoff', 'win_xsize', 'win_ysize'
+        that have the same meaning as in pygeoprocessing.iterblocks.
+    """
+    for yoff, ysize in zip(yoffs, ysizes):
+        for xoff, xsize in zip(xoffs, xsizes):
+            yield {
+                'xoff': xoff, 
+                'yoff': yoff,
+                'win_xsize': xsize,
+                'win_ysize': ysize}
+
+def random_array(shape, low=0, high=1, nodata=None, p_nodata=0.2, precision=2):
+    """Generate a random array useful as made-up raster data.
+
+    Args:
+        shape (tuple(int)): the shape of the array to make
+        low (float): the minimum possible value to include
+        high (float): the maximum possible value to include
+        nodata (int): If a nodata value is given, set some random elements
+            to this nodata value that may be outside the range.
+        p_nodata (float): A value in the range [0, 1] representing the
+            fraction of elements to make nodata (if a nodata value is provided)
+        precision (int): The number of decimal places to include
+
+    Returns:
+        numpy.ndarray with the given shape
+    """
+    magnitude = 10**precision
+    # multiplying then dividing by the magnitude gives the desired precision
+    array = numpy.random.randint(
+        low * magnitude, (high + 1) * magnitude, shape) / magnitude
+
+    if nodata is not None:  # could be zero
+        # randomly assign some values to nodata
+        nodata_indices = numpy.random.choice(
+            [False, True], shape, p=[(1 - p_nodata), p_nodata])
+        array[nodata_indices] = nodata
+    return array
+
+
+class StormwaterTests(unittest.TestCase):
```

Comment:
The random generation of test data in this class is an interesting idea!  While I like that this approach would ensure that we'll try to run the test suite on a large number of sample values, the fact that we can't reproduce issues in the tests if we _do_ have an error is worrisome.  Even if we wanted to export the seed, the Mersenne Twister has _so many_ possible states that exporting the current seed for numpy returns a multi-hundred element array that does have enough bits to represent the twister's seed.

It also seems like the math we're trying to check over is mostly pixel arithmetic, and so it makes me wonder if we're really gaining anything by trying lots of different numbers, with the cost being that failing tests are harder to debug, harder to reproduce, harder to implement and harder to maintain due to the duplicated math.  And with another variable, can we really rely on the tests to tell us when something is truly broken?

I know statically-defining a numpy array in source code is frustrating, but it does enable us to clearly see what values we expect.

Reply:
You make good points! To play devil's advocate, let me try to explain what I was thinking 😈  Can we really rely on the tests, when our expected result was generated by the same code that we're testing? I've encountered this a couple times at least (#492, #532) where we're testing against incorrect data. It's often not practical to manually compute more than one or two test data points, so we end up saving the model output as test data, perhaps spot-checking a few pixels, and leave it at that.

I guess there's really two components to this different test structure I'm trying out:
1. Randomly generating the input data
    - Pros: Cover a wider range of inputs, combinations of inputs
    - Cons: Difficult to save the information necessary to reproduce an error, thus harder to debug
2. Asserting that a mathematical relationship between the input and output holds true, rather than asserting that the actual output matches a predefined, expected output
    - Pros: Reduce hidden bugs where we are testing against incorrect data, fewer "magic numbers", being explicit about the relationship between input and output
    - Cons: More code to maintain, risk of duplicating the same error in source and test

(2) is necessary if you have (1), but you could use (2) without (1).
I may have overdone it here but my goal was to: *do the same calculation in the test, but in a different (preferably more explicit) way*. We can use the more efficient way in the source (e.g. array math) and the more explicit way in the test (e.g. pixel iteration). While it doesn't eliminate the possibility of error, it's much less likely that two different implementations would fail in the same way.

---------------------------------------------

Even though the distance algorithm is going away, I thought that test was a good example of a serious use case for (1) and (2). I wasn't super confident my algorithm would handle every case and it wasn't practical to manually check the distance raster for a complex multi-linestring.
- (1) made me feel more confident in the algorithm because I worried that some bizarre linestring shape would behave unexpectedly. Showing that it works on an arbitrary linestring vector seemed important given that it's not a 1-to-1 pixelwise operation.
- (2) was helpful because the optimization made it even harder to manually check the output. Doing the same distance calculation, without that optimization, verified that it wasn't missing anything.

Even though it made the test more complicated, I really don't know how else I would've made test data that I had confidence in. I would need to use that distance algorithm anyway to make the "expected" output, so why not be explicit about that and do it right in the test?

-----------------------------------------------
After considering this more, here's my revised thoughts on the subject. Would love to discuss more!

- (1) is _not_ very useful for 1-to-1 pixelwise algorithms because we can reasonably include a wide enough range of inputs in a static test case.
- (1) would be useful for testing things that aren't 1-to-1 pixelwise algorithms because it improves the chances of catching an unexpected bug arising from an "interaction" between inputs. This is only if we can find a way to save the problematic data so it's reproducible.
- (2) would be useful for a lot of things, but if it's so simple that it feels like you're just duplicating the same math, it's pointless.
- Whether using static or generated test cases, should be sure to include something from the whole range of possible values. Can it be a float? negative? zero? nodata? larger than 8, 16, 32, or 64 bit? If so, those should be represented.

Reply:
> Even though the distance algorithm is going away, I thought that test was a good example of a serious use case for (1) and (2). I wasn't super confident my algorithm would handle every case and it wasn't practical to manually check the distance raster for a complex multi-linestring.

Ah, this is a really interesting algorithm, and I do agree with you that this is an interesting use case for using a degree of randomness in the inputs and for asserting that the optimized algorithm produces expected results.  In this case, this test is serving as a test-driven development tool as much as it is a regression test, and I do think this works because of the relative simplicity with which the idea can be expressed in a less-than-optimal form for the sake of checking correctness.  So for this case, I think you're right, this sounds like a sensible and useful test!

But I think there are limits to the usefulness of duplicating the functionality.  In the case of per-pixel math, we're usually not doing anything fancy enough in a `raster_calculator` op to warrant duplicating the math in a test.  On the other end of the spectrum (let's say MFD routing), walking the graph and analyzing flow direction would be a pretty nontrivial operation to duplicate within a test on any meaningful scale.  The distance algorithm seems to occupy a sweet spot in between those two extremes.

So it seems like a possible rule of thumb is that this sort of testing might be practical for algorithms implemented in python, and perhaps not for cython-based algorithms?  In any case, the solution should always be appropriate for the problem at hand, and this test you're describing for this algorithm makes sense to me given what you're trying to accomplish.

> (1) is not very useful for 1-to-1 pixelwise algorithms because we can reasonably include a wide enough range of inputs in a static test case.

Yep, I agree!

> (1) would be useful for testing things that aren't 1-to-1 pixelwise algorithms because it improves the chances of catching an unexpected bug arising from an "interaction" between inputs. This is only if we can find a way to save the problematic data so it's reproducible.

In general, I think I agree with this provided that the randomness is also sensible for the inputs.  A DEM, for example, should exhibit DEM-like behavior (and there are intelligent ways of generating things like this).  If that's randomly generated AND we can reproduce it if there's a test failure, then this could be a good addition to the test suite.

I think there's a broader point here that you're getting at, which is that the tests themselves can be treated as a part of development, and if it's there and works _and_ is helping to explore the problem space, why not keep it?

But also yes, if we do have a degree of randomness, we'll want to be able to reproduce any failing results so that we can fix them.

> (2) would be useful for a lot of things, but if it's so simple that it feels like you're just duplicating the exact same math, it's pointless.

I'm not sure about pointless, but the duplication might outweigh the benefit :)

> Whether using static or generated test cases, should be sure to include something from the whole range of possible values. Can it be a float? negative? zero? nodata? larger than 8, 16, or 32 bit? If so, those should be represented.

This makes sense to me as a part of the development process for sure.  Rolling these sorts of checks into tests could also make sense, even if we haven't done this before.

Out of curiosity, I thought I'd look up what _The Pragmatic Programmer_ suggests: "property-based testing".  The python library [Hypothesis](https://hypothesis.readthedocs.io/en/latest/index.html) is an example of this and might be useful to look into!

Reply:
Thanks for making that ticket (#536) for this! I agree it's worth looking into outside of this PR. I removed all randomness from the stormwater tests and I left in some of the math-based testing (let me know if you think those should be removed too)

Reply:
Sure, that makes sense!